modules/feature-management/src/main/java/org/mule/runtime/feature/internal/togglz/provider/DefaultMuleTogglzFeatureProvider.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.feature.internal.togglz.provider;
+
+import static java.lang.String.format;
+import static java.util.Arrays.asList;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import org.mule.runtime.api.profiling.type.ProfilingEventType;
+import org.mule.runtime.feature.internal.togglz.MuleTogglzFeatureMetadata;
+import org.mule.runtime.feature.internal.togglz.MuleTogglzRuntimeFeature;
+import org.togglz.core.Feature;
+import org.togglz.core.metadata.FeatureMetaData;
+import org.togglz.core.metadata.enums.EnumFeatureMetaData;
+import org.togglz.core.spi.FeatureProvider;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Implementation of {@link FeatureProvider} for the Mule Runtime.
+ *
+ * @since 4.5.0
+ */
+public class DefaultMuleTogglzFeatureProvider implements MuleTogglzFeatureProvider {
+
+  public static final String RUNTIME_FEATURE_NOT_REGISTERED = "Runtime feature %s not registered";
+  public static final String FEATURE_ENUMS_NOT_NULL = "The featureEnums argument must not be null";
+  public static final String ARGUMENT_MUST_BE_ENUM = "The featureEnum argument must be an enum";
+  public static final String FEATURE_HAS_ALREADY_BEEN_ADDED = "The '%s' feature has already been added";
+  public static final String NO_FEATURES_ADDED = "There are no features added in this provider instance.";
+  public static final String CONSUMER_NAME_MUST_NOT_BE_NULL = "Consumer name must not be null.";
+  public static final String FEATURE_NAME_MUST_NOT_BE_NULL = "Feature name must not be null.";
+
+  private final Cache<org.mule.runtime.api.config.Feature, MuleTogglzRuntimeFeature> runtimeFeaturesCache =
+      Caffeine.newBuilder().build();
+
+  private Map<String, FeatureMetaData> metaDataCache = null;
```

Comment:
what about concurrent access to this map?

Reply:
added a ConcurrentHashMap